File: ATM.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Scanner;

class ATM 
{
    HashMap<String, String> account_details = new HashMap<>();
    HashMap<String, Double> balance_details = new HashMap<>();
    ArrayList<String> transaction_history = new ArrayList<>();
    Scanner sc = new Scanner(System.in);
    
    public void initialize_accounts() 
    {
        account_details.put("12345", "xyz");
        balance_details.put("12345", 15000.0);
        
        account_details.put("67890", "abc");
        balance_details.put("67890", 23000.0);
    }
    
    public boolean login_account(String account_number, String password) 
    {
        if(account_details.containsKey(account_number) && account_details.get(account_number).equals(password)) 
        {
            System.out.println("Login successful!");
            return true;
        } 
        else 
        {
            System.out.println("Invalid account number or password!");
            return false;
        }
    }
    
    public void check_balance(String account_number) 
    {
        double balance = balance_details.get(account_number);
        System.out.println("Your current balance is: " + balance);
        transaction_history.add("Checked balance: " + balance);
    }
    
    public void deposit(String account_number, double amount) {
        if (amount > 0) {
            double current_balance = balance_details.get(account_number);
            balance_details.put(account_number, current_balance + amount);
            System.out.println("Deposit successful! \n New balance: " + balance_details.get(account_number));
            transaction_history.add("Deposited: " + amount);
        } 
        else 
        {
            System.out.println("Invalid deposit amount!");
        }
    }
    
    public void withdraw(String account_number, double amount) {
        double current_balance = balance_details.get(account_number);
        if (amount > 0 && current_balance >= amount) {
            balance_details.put(account_number, current_balance - amount);
            System.out.println("Withdrawal successful! \n New balance: " + balance_details.get(account_number));
            transaction_history.add("Withdrew: " + amount);
        } 
        else 
        {
            System.out.println("Insufficient balance or invalid amount!");
        }
    }
    
    public void view_transaction_history() 
    {
        System.out.println("Transaction history:");
        for (String transaction : transaction_history) {
            System.out.println(transaction);
        }
    }

    public static void main(String[] args) {
        ATM atm = new ATM();
        atm.initialize_accounts(); 
        Scanner sc = new Scanner(System.in);
        
        System.out.println("Welcome to the ATM!!!");
        System.out.print("Enter account number: ");
        String account_number = sc.nextLine();
        
        System.out.print("Enter password: ");
        String password = sc.nextLine();
        
        if (atm.login_account(account_number, password)) {
            boolean exit = false;
            while(!exit) 
            {
                System.out.println("\nATM Menu:");
                System.out.println("1. Check Balance");
                System.out.println("2. Deposit Money");
                System.out.println("3. Withdraw Money");
                System.out.println("4. View Transaction History");
                System.out.println("5. Exit");
                
                System.out.print("Choose an option: ");
                int option = sc.nextInt();
                
                switch(option) 
                {
                    case 1:
                        atm.check_balance(account_number);
                        break;
                    case 2:
                        System.out.print("Enter amount to deposit: ");
                        double deposit_amount = sc.nextDouble();
                        atm.deposit(account_number, deposit_amount);
                        break;
                    case 3:
                        System.out.print("Enter amount to withdraw: ");
                        double withdraw_amount = sc.nextDouble();
                        atm.withdraw(account_number, withdraw_amount);
                        break;
                    case 4:
                        atm.view_transaction_history();
                        break;
                    case 5:
                        exit = true;
                        System.out.println("Thank you for using our ATM!!!");
                        break;
                    default:
                        System.out.println("Invalid option, Please try again!");
                        break;
                }
            }
        }
    }
}

File: ATM.c
#include<stdio.h>
#include<string.h>
#define Max_account 2
#define Max_history 10

struct Account
{
    char account_number[20];
    char password[8];
    double balance;
};

struct ATM
{
    struct Account accounts[Max_account];
    char transaction_history[Max_history][50];
    int transaction_count;
}atm;

void initialize_account(struct ATM *atm)
{
    strcpy(atm -> accounts[0].account_number, "12345");
    strcpy(atm -> accounts[0].password, "abc");
    atm -> accounts[0].balance = 10000;

    strcpy(atm -> accounts[1].account_number, "67890");
    strcpy(atm -> accounts[1].password, "xyz");
    atm -> accounts[1].balance = 23000;
}

int login(struct ATM *atm, char *acc_num, char *pass)
{
    for(int i = 0; i < 2; i++)
    {
        if(strcmp(atm -> accounts[i].account_number, acc_num) == 0)
        {
            if(strcmp(atm -> accounts[i].password, pass) == 0)
            {
            printf("Login Successful!\n");
            return i;
            }
        }
    }
    printf("Invalid Credentials!");
    return -1;
}

void check_balance(struct ATM *atm, int acc_ind)
{
    printf("Your Balance is %.2lf\n", atm -> accounts[acc_ind].balance);
    if(atm -> transaction_count < Max_history)
    {
        snprintf(atm -> transaction_history[atm -> transaction_count], 50, "Checked Balance: %.2lf", atm -> accounts[acc_ind].balance);
        atm -> transaction_count++;
    }
}

void deposit_money(struct ATM *atm, double amt, int acc_index)
{
    if(amt > 0)
    {
        printf("Amount %.2lf has been deposited!\n", amt);
        atm -> accounts[acc_index].balance += amt;

        printf("Your Balance: %.2lf\n", atm -> accounts[acc_index].balance);

        if(atm -> transaction_count < Max_history)
        {
            snprintf(atm -> transaction_history[atm -> transaction_count], 50, "Deposited money: %.2lf\n", amt);
            atm -> transaction_count++;
        }
    }
    else
        printf("Invalid amount!\n");
}

void withdraw_money(struct ATM *atm, double amt, int acc_index)
{
    if(amt > 0 && atm -> accounts[acc_index].balance >= amt)
    {
        printf("Amount %.2lf has been withdrawn!\n", amt);
        atm -> accounts[acc_index].balance -= amt;

        printf("Your Balance: %.2lf\n", atm -> accounts[acc_index].balance);


        if(atm -> transaction_count < Max_history)
        {
            snprintf(atm -> transaction_history[atm -> transaction_count], 50, "Withdrew money: %.2lf\n", amt);
            atm -> transaction_count++;
        }
    }
    else
        printf("Invalid amount!\n");
}

void transaction_history(struct ATM *atm)
{
    printf("Transaction history: \n");
    for(int i = 0; i < atm -> transaction_count; i++)
        printf("%s\n", atm -> transaction_history[i]);
}

int main()
{
    atm.transaction_count = 0;
    initialize_account(&atm);

    char acc_num[20], pass[20];
    int acc_index, option;
    double amount;
    int exit = 0;

    printf("Welcome to our ATM!!!\n");
    printf("Enter your account number: "); 
    scanf("%s", acc_num);
    printf("Enter your password: ");
    scanf("%s", pass);

    acc_index = login(&atm, acc_num, pass);

    if(acc_index != -1)
    {
        while(!exit)
        {
            printf("\nATM Menu:\n");
            printf("1. Check Balance\n");
            printf("2. Deposit Money\n");
            printf("3. Withdraw Money\n");
            printf("4. View Transaction History\n");
            printf("5. Exit\n");
            printf("Choose an option: ");
            scanf("%d", &option);

            switch(option)
            {
                case 1:
                    check_balance(&atm, acc_index);
                    break;

                case 2:
                    printf("Enter amount to be deposited: ");
                    scanf("%lf", &amount);
                    deposit_money(&atm, amount, acc_index);
                    break;

                case 3:
                    printf("Enter amount to be withdrawn: ");
                    scanf("%lf", &amount);
                    withdraw_money(&atm, amount, acc_index);
                    break;

                case 4:
                    transaction_history(&atm);
                    break;
                case 5:
                    exit = 1;
                    break;
                default:
                    printf("Please enter valid option!\n");
            }
        }
        printf("Thank you for using our ATM!!!");
    }
}